src/shogun/distributions/classical/GaussianDistributionNew.cpp

```diff
@@ -0,0 +1,297 @@
+		SGVector<float64_t> cov, bool cov_is_cholesky) :
+				CProbabilityDistribution(mean.vlen)
+{
+	REQUIRE(mean.vlen==cov.vlen,
+			"Mean dimension (%d) and covariance dimension (%d) do not match\n",
+			mean.vlen, cov.vlen);
+
+	init();
+
+	m_mean=mean;
+
+	m_cov=cov.convert_to_matrix(cov,1,cov.vlen,false);
+
+	m_cov_type="Vector";
+}
+
+CGaussianDistributionNew::CGaussianDistributionNew(SGVector<float64_t> mean,
+		SGMatrix<float64_t> cov, bool cov_is_cholesky) :
+				CProbabilityDistribution(mean.vlen)
+{
+	REQUIRE(cov.num_rows==cov.num_cols, "Covariance must be square but is "
+			"%dx%d\n", cov.num_rows, cov.num_cols);
+	REQUIRE(mean.vlen==cov.num_cols,
+			"Mean dimension (%d) and covariance dimension (%d) do not match\n",
+			mean.vlen, cov.num_cols);
+
+	init();
+
+	m_mean=mean;
+
+	if (!cov_is_cholesky)
+	{
+		compute_cholesky(cov);
+	}
+	else
+		m_cov=cov;
+
+	m_cov_type="Matrix";
+}
+
+CGaussianDistributionNew::~CGaussianDistributionNew()
+{
+
+}
+
+void CGaussianDistributionNew::set_mean(SGVector<float64_t> mean)
+{
+	m_mean=mean;
+}
+
+void CGaussianDistributionNew::set_cov(float64_t cov)
+{
+	m_cov=SGMatrix<float64_t>(1,1);
+	m_cov(0,0)=cov;
+
+	m_cov_type="Float";
+}
+
+void CGaussianDistributionNew::set_cov(SGVector<float64_t> cov)
+{
+	m_cov=cov.convert_to_matrix(cov,1,cov.vlen,false);
+
+	m_cov_type="Vector";
+}
+
+void CGaussianDistributionNew::set_cov(SGMatrix<float64_t> cov)
+{
+	m_cov=cov;
+
+	m_cov_type="Matrix";
+}
+
+SGVector<float64_t> CGaussianDistributionNew::get_mean() const
+{
+	return m_mean;
+}
+
+float64_t CGaussianDistributionNew::get_cov_spherical() const
+{
+	return *m_cov.data();
+}
+
+SGVector<float64_t> CGaussianDistributionNew::get_cov_diag() const
+{
+	return m_cov.get_row_vector(0);
+}
+
+SGMatrix<float64_t> CGaussianDistributionNew::get_cov_full() const
+{
+	return m_cov;
+}
+
+char* CGaussianDistributionNew::get_cov_type() const
+{
+	return m_cov_type;
+}
+
+SGMatrix<float64_t> CGaussianDistributionNew::sample(int32_t num_samples,
+		SGMatrix<float64_t> pre_samples) const
+{
+	REQUIRE(num_samples>0, "Number of samples (%d) must be positive\n",
+			num_samples);
+
+	/* use pre-allocated samples? */
+	SGMatrix<float64_t> samples;
+	if (pre_samples.matrix)
+	{
+		REQUIRE(pre_samples.num_rows==m_dimension, "Dimension of pre-samples"
+				" (%d) does not match dimension of Gaussian (%d)\n",
+				pre_samples.num_rows, m_dimension);
+
+		REQUIRE(pre_samples.num_cols==num_samples, "Number of pre-samples"
+				" (%d) does not desired number of samples (%d)\n",
+				pre_samples.num_cols, num_samples);
+
+		samples=pre_samples;
+	}
+	else
+	{
+		/* allocate memory and sample from std normal */
+		samples=SGMatrix<float64_t>(m_dimension, num_samples);
+		for (index_t i=0; i<m_dimension*num_samples; ++i)
+			samples.matrix[i]=sg_rand->std_normal_distrib();
+	}
+
+	/* map into desired Gaussian covariance */
+	Map<MatrixXd> eigen_samples(samples.matrix, samples.num_rows,
```

Comment:
This is the way to sample if m_cov_type == COV_FULL.
The other  covariance types work differently, so you need to do a switch(m_cov_type) here and impleent the corresponding behaviour (which is computationally cheaper)